KIPs/kip-111.md

```diff
@@ -0,0 +1,387 @@
+그저 카운터라면 4바이트이거나 8바이트등이 될 수도 있었을 텐데 왜 유독 7바이트가 되었을까? 그 이유는 ExtHash는 state trie, storage trie의 데이터에서 대부분의 데이터를 차지 하기 때문이다. 증가된 바이트 비율만큼 DB사이즈가 증가한다. 따라서 중복이 일어나지 않으면서 제일 작은 값을 찾아야 한다.  
+
+#### 1초에 업데이트 되는 얼마나 될까?
+현재 klaytn은 초당 4000 TPS를 처리할 수 있다. smart contract는 1번에 여러 구좌를 업데이트 시켜서 많은 노드 정보를 업데이트 시킬 수도 있다. 따라서 현실적으로 일어나기 어려울 정도로 큰 숫자로 정의해야 한다. 그래서 정한 숫자는 1블럭당 trie node에서 변경되는 노드수는 100만개 이하라고 가정했다.  
+ExtHashNonce는 어떻게 관리를 해야할까? 매번 업데이트 될때마다 DB에 넣는 것은 좋은 선택이 아니라고 생각했다. 그래서 시스템의 nano timestamp를 이용하기로 했다. 서버노드가 시작 될때 nano timestamp를 얻고 1바이트만큼 오르쪽으로 shift 연산을 한 값을 ExtHashNonce의 초기 값으로 잡고 새로운 노드가 생성될 때 마다 1씩 증가 시키기로 했다. 이러면 ExtHashNonce의 초기값이 중복될 일은 없다. nano timestamp를 이용한 ExtHashNonce는 1초 동안 몇개의 블록을 담을 수 있는지 계산해 보면
+```
+ nano timestamp의 1초는    : 1000,000,000
+ 1 byte shift right        : 1000,000,000 / 256 = 3906250 ( 7 byte )
+ 1 byte shift right 1 more : 1000,000,000 / 256 / 256 = 15258 ( 6 byte )
+```
+ExtHashNonce를 6 byte로 정할 경우 초당 15000개 이상의 trie-node가 업데이트되면 count 역전현상이 일어날 수 있고, 시스템이 재부팅 되면 ExtHashNonce는 충돌이 날 수 있다. ExtHashNonce가 7 byte 가 되면 초당 390만개 이상의 trie-node가 업데이트 되어야 하기때문에 충분히 여유로운 숫자이다.
+
+
+OR
+
+그럼 100년 동안 중복이 일어나지 않으려먼 몇 바이트가 필요할지 계산해 보자.
+```
+ 1 초   : 1000000
+ 1 시간 : 1000000 * 3600 ( 1시간은 3600초 )
+ 하루   : 1000000 * 3600 * 24 ( 하루는 24시간 )
+ 1년    : 1000000 * 3600 * 24 * 365 ( 1년은 365일 )
+ 100년  : 1000000 * 3600 * 24 * 365 * 100 ( 100년 )
+ 결과는 = 3153600000000000
+ 6 byte : 281474976710656
+ 7 byte : 72057594037927936
+```
+6바이트로는 부족하고 7바이트로는 충분하다.  
+
+
+ 
+
+초당 생성되는 trie의 노드수는 얼마정도로 봐야 할까? 현재 klaytn은 4000TPS를 처리할 수 있다. 한 구좌의 정보가 바뀌면 평균적으로 7~8개의 노드 정보가 업데이트 된다. smart contrcat는 동시에 여러 구좌의 정보가 업데이트될 수 있기 때문에 더 많은 노드의 정보가 업데이트 된다.  
+그럼 1 블럭이 생성될때 업데이트 되는 노드수는 어떠헥 잡아야 할까? 현실적으로  
+
+### Terminology
+
+Some borrowing of terminology, etc. - Beigepaper : An Ethereum Technical Specification. Author : Micah Daemeron
+
+a. Root Node - The Top (first) node in a tree
+
+b. Child Node - A node Directly connected to an-other node when moving away from the Root.
+
+c. Leaf Node - A node with no children
+
+d. Server Node - It means a server in charge of each role such as Consensus, Governance, Verification, and API in the klaytn network. CN, PN, and EN, which will be described later, are sub-concepts.
+
+e. CN - Consensus Node - Block을 생성하고 합의하고 검증하는 역할을 담당한다. CN노드는 CN노드들 끼리만 연결하며, 외부와의 연결을 PN노드가 대행한다.
+
+f. PN - Proxy Node - CN노드가 CN노드끼리의 통신에 집중할 수 있도록 외부와의 통신을 대행해주고 CN의 백업서버 역할을 담당한다.
+
+g. EN - End Node - 여러 형태의 외부 API요청을 처리하여 CN, PN의 부하를 감소시킬 목적의 서버
+
+h. MPT - Merkle Patrica Trie
+
+
+
+### Trie 구조 정의
+ExtHash는 State Trie Node에서만 사용한다. Header, Body, Transaction, Reciept등은 중복문제가 발생하지 않기 때문에 32byte hash를 사용한다.
+* State Trie Node는 32byte Hash와 39byte ExtHash 중 1가지만 사용된다.
+* 기존과 동일하게 동작하는 Server Node에서는 32byte Hash를 사용한다.
+* statedb pruning을 사용하는 Server Node에서는 39byte Hash를 사용한다.
+```
+ref_32    = 32 byte - Hash
+ref_39    = 39 byte - ExtHash
```

Comment:
`ExtHash = Hash | ExtHashNonce` right ? instead of `ExtHash = ExtHashNonce | Hash`
(`|`: concat)

Reply:
add comment to 

ref_32    = 32 byte - Hash
ref_39    = 39 byte - ExtHash = Hash(32 byte) + ExtHashNonce(7 byte)